ml-agents/mlagents/trainers/ppo/multi_gpu_policy.py

```diff
@@ -0,0 +1,187 @@
+import logging
+import numpy as np
+
+from mlagents.envs.timers import timed
+from mlagents.trainers.ppo.policy import PPOPolicy
+from mlagents.trainers.components.reward_signals.reward_signal_factory import (
+    create_reward_signal,
+)
+from mlagents.trainers.components.bc.module import BCModule
+
+logger = logging.getLogger("mlagents.trainers")
+
+
+class MultiGPUPPOPolicy(PPOPolicy):
+    def __init__(self, seed, brain, trainer_params, is_training, load):
+        reward_signal_configs = trainer_params["reward_signals"]
+        self.reward_signals = {}
+
+        self.devices = get_devices()
+        self.towers = []
+        
+        with self.graph.as_default():
+            for device in devices:
+                with tf.device(device):
+                    with tf.variable_scope(TOWER_SCOPE_NAME, reuse=tf.AUTO_REUSE):
+                        self.towers.append(PPOModel(
+                                brain=brain,
+                                lr=float(trainer_params["learning_rate"]),
+                                h_size=int(trainer_params["hidden_units"]),
+                                epsilon=float(trainer_params["epsilon"]),
+                                beta=float(trainer_params["beta"]),
+                                max_step=float(trainer_params["max_steps"]),
+                                normalize=trainer_params["normalize"],
+                                use_recurrent=trainer_params["use_recurrent"],
+                                num_layers=int(trainer_params["num_layers"]),
+                                m_size=self.m_size,
+                                seed=seed,
+                                stream_names=list(reward_signal_configs.keys()),
+                                vis_encode_type=trainer_params["vis_encode_type"],
+                            )
+                        )
+                        self.towers[-1].create_ppo_optimizer()
+
+            # Create reward signals
+            for reward_signal, config in reward_signal_configs.items():
```

Comment:
This reward signal logic and the pretrainer logic below is shared with the base class `__init__`.  It would probably be best to extract it somehow so we don't have to duplicate it.